config/src/iroha.rs

```diff
@@ -60,26 +62,82 @@ view! {
+            disable_panic_terminal_colors: Some(bool::default()),
+            kura: Some(kura::ConfigurationProxy::default()),
+            sumeragi: Some(sumeragi::ConfigurationProxy::default()),
+            torii: Some(torii::ConfigurationProxy::default()),
+            block_sync: Some(block_sync::ConfigurationProxy::default()),
+            queue: Some(queue::ConfigurationProxy::default()),
+            logger: Some(logger::ConfigurationProxy::default()),
+            genesis: Some(genesis::ConfigurationProxy::default()),
+            wsv: Some(wsv::ConfigurationProxy::default()),
+            network: Some(network::ConfigurationProxy::default()),
+            telemetry: Some(telemetry::ConfigurationProxy::default()),
+        }
+    }
+}
+
+impl ConfigurationProxy {
+    /// Finalisation of Iroha config, i.e. in case where only some of mutually equivalent fields
+    /// were provided, instantiation of all the missing fields happens.
+    ///
+    /// # Errors
+    /// Bails early if none of the fields from the corresponding mutually equivalent set were
+    /// provided.
+    #[allow(clippy::expect_used, clippy::unwrap_in_result)]
+    pub fn finalize(&mut self) -> Result<()> {
+        if let Some(sumeragi_proxy) = &mut self.sumeragi {
+            // First, iroha public/private key and sumeragi keypair are interchangeable, but
+            // the user is allowed to provide only the former, and keypair is generated automatically
+            if let (Some(public_key), Some(private_key)) = (&self.public_key, &self.private_key) {
+                sumeragi_proxy.key_pair =
+                    Some(KeyPair::new(public_key.clone(), private_key.clone())?);
+            } else {
+                eyre::bail!(ConfigError::ProxyBuildError(
+                    "Neither Iroha public and private key nor sumeragi keypair supplied".to_owned()
+                ))
+            }
+            // Second, torii gateway and sumeragi peer id are interchangeable too
+            if let Some(torii_proxy) = &mut self.torii {
+                if let Some(peer_id) = &sumeragi_proxy.peer_id {
+                    torii_proxy.p2p_addr = Some(peer_id.address.clone());
+                } else {
+                    sumeragi_proxy.peer_id = Some(iroha_data_model::peer::Id::new(
+                        &torii_proxy.p2p_addr.clone().ok_or_else(|| {
+                            eyre::eyre!("Torii `p2p_addr` field has `None` value")
+                        })?,
+                        &self.public_key.clone().expect(
+                            "Iroha `public_key` should have been initialized above at the latest",
+                        ),
+                    ));
+                }
+            }
+            // Finally, if trusted peers were not supplied, we can fall back to inserting itself as
+            // the only trusted one
+            if sumeragi_proxy.trusted_peers.is_none() {
+                sumeragi_proxy.insert_self_as_trusted_peers()
+            }
```

Comment:
Technically the `Configuration` would fail at build stage, as the `torii.p2p_addr` is `None` by default. We could bail early here, e.g. adding `else` after `if let Some(torii_proxy) = ..` in finalization, but I guess it would already be informative enough when one of the `expect()` calls fails or `build()` crashes.

Reply:
Just encountered that sort of error during property testing. I think the freshly bumped version of this func should do the trick now.